### ISSUE TITLE
Inconsistent/Incompatible handling of filename escaping in multipart/form-data compared to RFC 7578 and browsers

### ISSUE THREAD

**Issue by sleevi** [curl/curl#7789]
### I did this

1. Create a file of the form `"foo.jpg` (note the leading quote)
2. Submit as a form using `curl -i -F encoded_image=@\"foo.jpg https://www.google.com/searchbyimage/upload`

### I expected the following

Within the `multipart/form-data` that is generated, the `filename` parameter of the request will be in the following form: `filename="\"foo.jpg"` - using the `quoted-string` production of RFC 822.

However, I would expect that the filename would of the form `filename="%22foo.jpg"`, using the [Percent-Encoding Option of RFC 7578](https://datatracker.ietf.org/doc/html/rfc7578#section-2)

### curl/libcurl version

ToT

### Further Details

The context here is that `multipart/form-data` encoding is handled by [lib/formdata.c](https://github.com/curl/curl/blob/e081048c44ec446bf6213d6b45a25ca51fd8a59d/lib/formdata.c). The form-data support is implemented using the generic MIME encoder, in the family of `curl_mime_*` functions. For example, in the above example of attaching a file, this is handled by https://github.com/curl/curl/blob/e081048c44ec446bf6213d6b45a25ca51fd8a59d/lib/formdata.c#L882 calling [`curl_mime_filedata`](https://github.com/curl/curl/blob/52fab72397687467650093c86e5479cb1d759042/lib/mime.c#L1435-L1482)

`curl_mime_filedata` passes the filename onwards using [`curl_mime_filename`](https://github.com/curl/curl/blob/52fab72397687467650093c86e5479cb1d759042/lib/mime.c#L1384-L1400), passing only the basename of the filename.

Later, when compiling the multipart message, the field name and filename for the `Content-Disposition` header are escaped using the [`escape_string` function](https://github.com/curl/curl/blob/52fab72397687467650093c86e5479cb1d759042/lib/mime.c#L281-L305), as shown at https://github.com/curl/curl/blob/52fab72397687467650093c86e5479cb1d759042/lib/mime.c#L1867-L1886

These is a complex area due to issues of character sets and filenames. Within this space, [RFC 6266, Section 4.3](https://datatracker.ietf.org/doc/html/rfc6266#section-4.3) is relevant, and further expanded upon in [RFC 6266, Appendix C.2](https://datatracker.ietf.org/doc/html/rfc6266#appendix-C.2). [RFC 7578, Section 4.2](https://datatracker.ietf.org/doc/html/rfc7578#section-4.2) attempts to provide guidance here with respect to the `filename` handling inter-operably within a `Content-Disposition: form-data` part.

The [HTML Living Standard Multipart form data](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#multipart-form-data) specification places a normative dependency on RFC 7578, providing further guidance with respect to the encoding and escaping of `field names` and `filenames for file fields`, the two areas escaped via `escape_string`, to encode to the appropriate encoding and then replace `{ 0x0A, 0X0D, 0X22 }` with `{ "%0A", "%0D", "%22" }`, respectively. This is compatible with the guidance of RFC 7578, and reflects widespread use within browsers (e.g. [Chromium/Chrome](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/network/form_data_encoder.cc;l=53-77;drc=4cd749d0d82138ff31ed3a2bc5d925bb6d83fe16) or [WebKit/Safari](https://github.com/WebKit/WebKit/blob/e2db8292bebd8c89d8e02c54b6b5d5a3b0757596/Source/WebCore/platform/network/FormDataBuilder.cpp#L61-L85))

This is an area where MIME attachments differ in practice than form data. It seems that it may be useful to at least align the form encoder to match the approach mentioned within RFC 7578, for greater harmonization.

**Comment by bagder** [comment]
I agree. @monnerat, do you want to grab this?

**Comment by monnerat** [comment]
This is a picky area: we have something that seems to work for forms and featuring mime did not change this escaping that was already present in original formdata code.

In addition, there are contradictory informations in the mentioned documents:

In https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#multipart-form-data:
> For field names and filenames for file fields, the result of the encoding in the previous bullet point must be escaped by replacing any 0x0A (LF) bytes with the byte sequence `%0A`, 0x0D (CR) with `%0D` and 0x22 (") with `%22`. The user agent must not perform any other escapes.

In https://datatracker.ietf.org/doc/html/rfc6266#section-4.3:
> 
> Note: Many user agents do not properly handle the escape character
>       "\" when using the quoted-string form.  Furthermore, some user
>       agents erroneously try to perform unescaping of "percent" escapes
>       (see Appendix C.2), and thus might misinterpret filenames
>       containing the percent character followed by two hex digits.

So: %-escape or not, and what about the backslash ???

I fear this change may break a lot of things: in particular, tests 39 and 1158.
I'm not convinced it is wise to introduce problems where there is no bug, just because of an inconsistency with third party applications.

As a consequence, I can do it, but things have to be clarified first.

**Comment by bagder** [comment]
This is rather tricky. It seems the three specs have very different ideas on how to encode some basic bytes, like for example a CR and a double-quote:

| ascii  | RFC6266 | RFC 7578 | WHATWG |
|----|----|----|---|
| 0x0d | unclear | %0d | %0d |
| 0x22 (")  | \\" | %22 | " |

Or did I get any of them wrong?


> I fear this change may break a lot of things: in particular, tests 39 and 1158.

Those tests verify curl's behavior. If we change behavior, we of course also need to update those tests accordingly!

When specs are hard to follow (like here) I think we need to check what other tools and libraries use and what widely deployed server-side applications accept/support. I think unifying with browser behavior on this seems reasonable as they are probably the major producers of formdata bodies.

**Comment by sleevi** [comment]
> So: %-escape or not, and what about the backslash ???

So this isn’t quite the [problem](https://daniel.haxx.se/blog/2017/01/30/one-url-standard-please/) with [URLs](https://daniel.haxx.se/blog/2016/05/11/my-url-isnt-your-url/), but similar.

In order of precedence, and recognizing that cURL may ignore the WHATWG work due to the Living Standard nature:
* WHATWG HTML Living Standard
* RFC 7578
* RFC 6266 (ish)

I say “ish”, because _virtually_ no one is using the `paramname*` construction to extend `paramname` with character set details.

> It seems the three specs have very different ideas on how to encode some basic bytes.

That table isn’t correct. WHATWG is fully compatible with 7578, and both encode `DQUOTE` as `%22` (i.e. this issue). Where they differ is representing characters outside of 7-bit ASCII: WHATWG normalizes encoding sets and then converts those out of range to entities, while 7578 doesn’t normalize and just entitizes, at least, AFAICT.

**Comment by bagder** [comment]
Yes sorry, I realized my mistake and edited the table. WHATWG and RFC 7578 agree on 0x0a, 0x0d and 0x22.

I think I'm wrong with the backslash there as 7578 doesn't say it should be encoded I think...

**Comment by sleevi** [comment]
@bagder I think your table is correct now, in that Section 2 of RFC 7578 says you can percent encode the `disallowed` characters - which naively would include such.

However, I don’t think `\` is necessarily an interesting case, primarily because I understand browsers treat it as a path separator, and thus it’s elided from the filename before encoding, similar to how cURL grabs the basename of the file. That said, I could see it being entirely reasonable for WHATWG to make that a bit more explicit, either in processing or making sure it’s addressed for encoding (to avoid mis-escaped sequences, similar to the `”` handling)

**Comment by monnerat** [comment]
> Or did I get any of them wrong?

Seems not. But you know http better than I do.

> Those tests verify curl's behavior. If we change behavior, we of course also need to update those tests accordingly!

Sure! But this shows we have no bug, but perhaps a misinterpretation. That said, if tests break, applications running perfectly for more that 13 years now may fail too.

> ... and what widely deployed server-side applications accept/support.

I think it's the most important. Servers themselves may parse it too (Apache modules?) As we only generate these and do not parse them, It would be good to know how widespread CGI-like apps and servers (i.e.: PHP) understand them.

> I say “ish”, because virtually no one is using the paramname* construction to extend paramname with character set details.

Curl does not support charset encoding in http/mime header fields. This would be rather cumbersome to implement.

> I think I'm still wrong with the backslash there as 7578 doesn't say it should be encoded I think...

AFAICS, you're as puzzled as I am :-)

As the various documents are not OK with each other, this introduces a huge spelunking work for us in third-party products. Even the latter may not all stick to the same behavior. I'm not sure I'm ready for thst :-(

**Comment by sleevi** [comment]
> this introduces a huge spelunking work for us in third-party products.

Indeed! I mentioned separately to @bagder that this began as an “Oh, that’s odd” when someone noticed a webpage behaving differently if browsers sent the request vs cURL, and the deeper I went, the more boxes of spiders I found myself opening.

I thought it useful to write this down and try to capture the references. cURL’s ubiquity suggests this would be the main place people run into it.

From a browser evolution space, there’s been a wide variety of behaviours before converging on the current approach. For example, one was just to replace the `”` with a `?` (as with any other high-bit character), another is to ignore the input filename and generate a sequential or random filename if it couldn’t be represented.

Although ultimately this is only intended as a hint for servers (for the reasons and risks noted in the various specs), and to that end, many behaviors are valid, I figured I would at least file this so it’s on your radar, even if the risk/complexity/ineffability is too much :)

**Comment by monnerat** [comment]
> Although ultimately this is only intended as a hint for servers ...

Before speaking semantics, we should care about lexical and syntax. If we do not encode a backslash, would it be correct or would it be interpreted as a quoted-pair (http uses non-terminal `quoted-string` while it is defined in RFC 822!). Furthermore, from what I understand from the documents, this should be applied to field names too (although non-ascii is discouraged for them). And those are not just hints...

**Comment by bagder** [comment]
Firefox 88 on Linux for the file name `"foo`:
~~~
-----------------------------104053578414215370022244339964
Content-Disposition: form-data; name="secret"; filename="\"foo"
Content-Type: application/octet-stream

contents
~~~
Chrome 94 on the same box same name:
~~~
------WebKitFormBoundaryApcVGpeqb4k55wB8
Content-Disposition: form-data; name="secret"; filename="%22foo"
Content-Type: application/octet-stream

contents
~~~

**Comment by bagder** [comment]
It's even worse. I edited my html for the form to be:
~~~
  File: <input type="file" name="secret
"><br>
~~~
(the newline there is intended)

The results from the jury:

Firefox 88:
~~~
Content-Disposition: form-data; name="secret "; filename="\"foo"
~~~
Chrome 94:
~~~
Content-Disposition: form-data; name="secret%0D%0A"; filename="%22foo"
~~~

**Comment by monnerat** [comment]
I tried with firefox 92.0 and it uses percent encoding without backslash escaping.

**Comment by monnerat** [comment]
Maybe some M$ tester could check with IE and edge?

**Comment by bagder** [comment]
It still means that receivers of multipart formposts today have to deal with both the \\-escaped **and** percent encoded fields for now and the near-term future. Not sure that helps us with this issue, I'm just stating the obvious.

**Comment by monnerat** [comment]
Escaping-style detection (if some) is a server-side problem, no change is then required in curl for that. And sooner or later, if major clients move to %-escaping, receivers will have to adapt :-(
The way detection should be done is unfortunately not specified: this will guarantee here again that, even if some server-side escaping-style detection is implemented somewhere, there's very few chance that another application will have the same algorithm.

I checked PHP 7.4.24: it uses raw name values as `$_POST` and` $_FILES` keys and raw filename values as `$_FILES[]['name']`. The PHP scripts are thus responsible for unescaping them.

To be versatile on our side, I can only suggest we have an additional `CURLOPT_` for that (with which default?).
